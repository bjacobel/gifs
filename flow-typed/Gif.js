type Gif = {
  observedHeight: number,
  id: number,
  src: string,
};
