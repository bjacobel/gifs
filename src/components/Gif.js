import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

export default class Gif extends Component {
  componentWillMount() {
    const { watchForSize, id, name, image } = this.props;

    this.state = {
      loaded: false,
      loading: false
    };

    this.onImageLoad = () => {
      this.setState({
        loaded: true,
        loading: false
      });
      image.removeEventListener('load', this.onImageLoad);
    };
    this.startLoad = () => {
      this.state = Object.assign({}, this.state, { loading: true });
    };

    watchForSize(image, id, name, document.querySelector('.column.gifs').scrollWidth);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.image.src.startsWith('https://gifs.')) {
      if (nextProps.image.complete) {
        this.onImageLoad();
      } else {
        nextProps.image.addEventListener('load', this.onImageLoad);
      }
    }
  }

  render() {
    const { image } = this.props;
    const { loaded, loading } = this.state;
    return (
      <div>
        <div className={ classnames('loading-indicator', { loaded, loading }) } >
          <div className="spinner" />
        </div>
        <img src={ image.src } role="presentation" onMouseOver={ this.startLoad } />
      </div>
    );
  }
}

Gif.propTypes = {
  image: PropTypes.shape({
    src: PropTypes.string.isRequired
  }).isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  watchForSize: PropTypes.func.isRequired
};
