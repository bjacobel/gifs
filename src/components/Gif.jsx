import React, { Component, PropTypes } from 'react';

export default class Gif extends Component {
  // @TODO: handle .gifv, .webm, .mp4 (<video> tag)
  // also would be cool if it could not animate until mouseover
  // will probably require a lambda and imagemagick or pillow

  componentWillMount() {
    const { gif } = this.props;

    this.img = new Image();
    this.img.src = `https://gifs.bjacobel.com/${gif.src}`;
    this.img.addEventListener('load', () => {
      const canvas = document.getElementById(`canvas-${gif.id}`);
      const ctx = canvas.getContext('2d');
      ctx.drawImage(this.img, 0, 0);
    }, false);
  }

  render() {
    const { gif } = this.props;

    const showGif = () => {

    };

    const showCanvas = () => {

    };

    return (
      <div onMouseEnter={ showGif() } onMouseLeave={ showCanvas() }>
        <canvas id={ `canvas-${ gif.id }` } />
        <img src={ `https://gifs.bjacobel.com/${gif.src}` } />
      </div>
    );
  }
}

Gif.propTypes = {
  gif: PropTypes.shape({
    src: PropTypes.string.isRequired
  }).isRequired
};
