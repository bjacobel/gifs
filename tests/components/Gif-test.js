import React from 'react';
import { shallow } from 'enzyme';

import Gif from '../../src/components/Gif';

describe('Gif component', () => {
  let gif;
  let image;

  beforeEach(() => {
    document.querySelector = jest.fn(() => ({ scrollWidth: 500 }));
    image = new Image();
    image.src = 'https://gifthumbs.bjacobel.com/partyparrot.gif';

    gif = (
      <Gif
        image={ image }
        id="asdf"
        name="testgif"
        watchForSize={ jest.fn() }
      />
    );
  });

  it('matches the initial state snapshot', () => {
    expect(shallow(gif)).toMatchSnapshot();
  });

  it('matches snapshot after load of the full gif is triggered', () => {
    const renderedGif = shallow(gif);

    renderedGif.find('img').simulate('mouseOver');
    image.src = 'https://gifs.bjacobel.com/partyparrot.gif';
    renderedGif.setProps('image', image);

    expect(renderedGif).toMatchSnapshot();
  });

  it('matches snapshot after load of the full gif is finished', () => {
    const renderedGif = shallow(gif);

    renderedGif.find('img').simulate('mouseOver');
    image.src = 'https://gifs.bjacobel.com/partyparrot.gif';
    renderedGif.setProps('image', image);
    image.dispatchEvent(new Event('load'));
    renderedGif.update();

    expect(renderedGif).toMatchSnapshot();
  });

  it('unmounts and removes the event listener', () => {
    const renderedGif = shallow(gif);

    renderedGif.find('img').simulate('mouseOver');
    image.src = 'https://gifs.bjacobel.com/partyparrot.gif';
    renderedGif.setProps('image', image);
    image.dispatchEvent(new Event('load'));
    renderedGif.unmount();

    expect(image.src).toEqual('//:0');
  });
});
