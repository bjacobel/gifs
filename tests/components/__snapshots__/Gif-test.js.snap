exports[`Gif component matches snapshot after load of the full gif is finished 1`] = `
<div>
  <div
    className="loading-indicator loaded">
    <div
      className="spinner" />
  </div>
  <img
    onMouseOver={[Function]}
    role="presentation"
    src="https://gifs.bjacobel.com/partyparrot.gif" />
</div>
`;

exports[`Gif component matches snapshot after load of the full gif is triggered 1`] = `
<div>
  <div
    className="loading-indicator loading">
    <div
      className="spinner" />
  </div>
  <img
    onMouseOver={[Function]}
    role="presentation"
    src="https://gifs.bjacobel.com/partyparrot.gif" />
</div>
`;

exports[`Gif component matches the initial state snapshot 1`] = `
<div>
  <div
    className="loading-indicator">
    <div
      className="spinner" />
  </div>
  <img
    onMouseOver={[Function]}
    role="presentation"
    src="https://gifthumbs.bjacobel.com/partyparrot.gif" />
</div>
`;
